File: Python/오픈채팅방.py
def solution(record):
    trace = [(v.split()[0], v.split()[1]) for v in record if v.split()[0] != "Change"]
    ids = list(set([id.split()[1] for id in record]))
    nick = dict(zip(ids, '0' * len(ids)))
    for r in record:
        if r.split()[0] == "Leave":
            continue
        nick.update({r.split()[1]: r.split()[2]})

    return [nick[t[1]] + "님이 들어왔습니다." if t[0] == "Enter" else nick[t[1]] + "님이 나갔습니다." for t in trace]

File: Kotlin/오픈채팅방.kt
fun solution(record: Array<String>): Array<String> {

    val map = mutableMapOf<String, String>()
    val arr = mutableListOf<Array<String>>()

    record.forEach {
        val i = it.split(" ")
        when (i[0]) {
            "Enter" -> {
                map[i[1]] = i[2]
                arr.add(arrayOf(i[0], i[1]))
            }
            "Leave" -> {
                arr.add(arrayOf(i[0], i[1]))
            }
            "Change" -> {
                map[i[1]] = i[2]
            }
        }
    }
    return arr.map {
        if (it[0] == "Enter") {
            "${map[it[1]]}님이 들어왔습니다."
        } else {
            "${map[it[1]]}님이 나갔습니다."
        }
    }.toTypedArray()
}